File: ios/Infrastructure/MLKit/Factories/TextRecognitionServiceFactory.swift
import Foundation
import NitroModules

#if MLKIT_TEXT_RECOGNITION_ANY
  import MLKitTextRecognition
#endif

#if MLKIT_TEXT_RECOGNITION_CHINESE
  import MLKitTextRecognitionChinese
#endif

#if MLKIT_TEXT_RECOGNITION_DEVANAGARI
  import MLKitTextRecognitionDevanagari
#endif

#if MLKIT_TEXT_RECOGNITION_JAPANESE
  import MLKitTextRecognitionJapanese
#endif

#if MLKIT_TEXT_RECOGNITION_KOREAN
  import MLKitTextRecognitionKorean
#endif

#if MLKIT_TEXT_RECOGNITION_ANY
  class TextRecognitionServiceFactory {

    /// Recognizer options for `language`, or `nil` when that language model was
    /// excluded from this build by the selective ML Kit configuration.
    ///
    /// The `#if` guards live inside each `case` body rather than around the
    /// `case` labels themselves: that keeps the switch exhaustive under every
    /// combination of model flags, so adding a language to
    /// `DomainTextRecognitionLanguage` still fails the build until it is
    /// handled here.
    private static func makeOptions(for language: DomainTextRecognitionLanguage)
      -> CommonTextRecognizerOptions?
    {
      switch language {
      case .latin:
        #if MLKIT_TEXT_RECOGNITION
          return MLKitTextRecognition.TextRecognizerOptions()
        #else
          return nil
        #endif
      case .chinese:
        #if MLKIT_TEXT_RECOGNITION_CHINESE
          return ChineseTextRecognizerOptions()
        #else
          return nil
        #endif
      case .devanagari:
        #if MLKIT_TEXT_RECOGNITION_DEVANAGARI
          return DevanagariTextRecognizerOptions()
        #else
          return nil
        #endif
      case .japanese:
        #if MLKIT_TEXT_RECOGNITION_JAPANESE
          return JapaneseTextRecognizerOptions()
        #else
          return nil
        #endif
      case .korean:
        #if MLKIT_TEXT_RECOGNITION_KOREAN
          return KoreanTextRecognizerOptions()
        #else
          return nil
        #endif
      }
    }

    /// Configuration key that enables `language`'s model, as accepted by the
    /// `$VisionCameraMLKit` Podfile hash and the Expo config plugin.
    private static func configurationKey(for language: DomainTextRecognitionLanguage) -> String {
      switch language {
      case .latin: return "textRecognition"
      case .chinese: return "textRecognitionChinese"
      case .devanagari: return "textRecognitionDevanagari"
      case .japanese: return "textRecognitionJapanese"
      case .korean: return "textRecognitionKorean"
      }
    }

    static func createTextRecognizer(language: DomainTextRecognitionLanguage) throws
      -> TextRecognizer
    {
      guard let options = makeOptions(for: language) else {
        throw RuntimeError.error(
          withMessage:
            "Text recognition for \(language.rawValue) is not enabled in this build. "
            + "Set \(configurationKey(for: language)) to true in your ML Kit configuration."
        )
      }

      return TextRecognizer.textRecognizer(options: options)
    }

    static func create(language: DomainTextRecognitionLanguage = .latin) throws
      -> MLKitTextRecognitionService
    {
      let textRecognizer = try createTextRecognizer(language: language)
      return MLKitTextRecognitionService(textRecognizer: textRecognizer)
    }
  }
#endif  // MLKIT_TEXT_RECOGNITION_ANY

File: android/src/main/java/com/visioncameramlkit/infrastructure/mlkit/factories/TextRecognitionServiceFactory.kt
package com.visioncameramlkit.infrastructure.mlkit.factories

import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.TextRecognizer
import com.google.mlkit.vision.text.TextRecognizerOptionsInterface
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.visioncameramlkit.BuildConfig
import com.visioncameramlkit.domain.models.TextRecognitionLanguage
import com.visioncameramlkit.infrastructure.mlkit.MLKitTextRecognitionService

object TextRecognitionServiceFactory {
  @Suppress("LongMethod", "ThrowsCount")
  fun createTextRecognizer(language: TextRecognitionLanguage): TextRecognizer {
    val options: TextRecognizerOptionsInterface =
      when (language) {
        TextRecognitionLanguage.LATIN -> {
          @Suppress("KotlinConstantConditions")
          if (!BuildConfig.MLKIT_TEXT_RECOGNITION) {
            throw IllegalStateException(
              "Text recognition for Latin is not enabled. " +
                "Set textRecognition to true in your MLKit configuration.",
            )
          }
          TextRecognizerOptions.DEFAULT_OPTIONS
        }

        TextRecognitionLanguage.CHINESE -> {
          @Suppress("KotlinConstantConditions")
          if (!BuildConfig.MLKIT_TEXT_RECOGNITION_CHINESE) {
            throw IllegalStateException(
              "Text recognition for Chinese is not enabled. " +
                "Set textRecognitionChinese to true in your MLKit configuration.",
            )
          }
          createOptionsViaReflection(
            "com.google.mlkit.vision.text.chinese.ChineseTextRecognizerOptions",
          )
        }

        TextRecognitionLanguage.DEVANAGARI -> {
          @Suppress("KotlinConstantConditions")
          if (!BuildConfig.MLKIT_TEXT_RECOGNITION_DEVANAGARI) {
            throw IllegalStateException(
              "Text recognition for Devanagari is not enabled. " +
                "Set textRecognitionDevanagari to true in your MLKit configuration.",
            )
          }
          createOptionsViaReflection(
            "com.google.mlkit.vision.text.devanagari.DevanagariTextRecognizerOptions",
          )
        }

        TextRecognitionLanguage.JAPANESE -> {
          @Suppress("KotlinConstantConditions")
          if (!BuildConfig.MLKIT_TEXT_RECOGNITION_JAPANESE) {
            throw IllegalStateException(
              "Text recognition for Japanese is not enabled. " +
                "Set textRecognitionJapanese to true in your MLKit configuration.",
            )
          }
          createOptionsViaReflection(
            "com.google.mlkit.vision.text.japanese.JapaneseTextRecognizerOptions",
          )
        }

        TextRecognitionLanguage.KOREAN -> {
          @Suppress("KotlinConstantConditions")
          if (!BuildConfig.MLKIT_TEXT_RECOGNITION_KOREAN) {
            throw IllegalStateException(
              "Text recognition for Korean is not enabled. " +
                "Set textRecognitionKorean to true in your MLKit configuration.",
            )
          }
          createOptionsViaReflection(
            "com.google.mlkit.vision.text.korean.KoreanTextRecognizerOptions",
          )
        }
      }

    return TextRecognition.getClient(options)
  }

  /**
   * Creates language-specific TextRecognizerOptions via reflection to avoid hard
   * class dependencies on variant modules that may not be included in the build.
   * Each variant options class follows the Builder pattern: OptionsClass.Builder().build()
   */
  private fun createOptionsViaReflection(className: String): TextRecognizerOptionsInterface {
    try {
      val builderClass = Class.forName("$className\$Builder")
      val builder = builderClass.getDeclaredConstructor().newInstance()
      val buildMethod = builderClass.getMethod("build")
      return buildMethod.invoke(builder) as TextRecognizerOptionsInterface
    } catch (e: ClassNotFoundException) {
      throw IllegalStateException(
        "MLKit class $className not found. " +
          "Ensure the corresponding text recognition dependency is included.",
        e,
      )
    }
  }

  fun create(language: TextRecognitionLanguage = TextRecognitionLanguage.LATIN): MLKitTextRecognitionService {
    val textRecognizer = createTextRecognizer(language)
    return MLKitTextRecognitionService(textRecognizer)
  }
}
